expression/expression.go

```diff
@@ -173,6 +173,27 @@ func IsEQCondFromIn(expr Expression) bool {
 	return len(cols) > 0
 }
 
+// NotNull checks if an expression is constantly not null.
+func NotNull(l Expression) bool {
+	switch l := l.(type) {
+	case *Column:
+		return mysql.HasNotNullFlag(l.GetType().Flag)
+	case *Constant:
+		return !l.Value.IsNull()
+	case *CorrelatedColumn:
+		return mysql.HasNotNullFlag(l.GetType().Flag)
+	case *ScalarFunction:
+		for _, arg := range l.GetArgs() {
```

Comment:
Though the current implementation may satisfy the `InSubquery`'s needs. But for the real `NotNull` property of scalar function, we need to take the `IsNull` sig in to special consideration at least.

Reply:
yes, for example `arg is true` is not null when arg is null.